test/cpp/src/201-register-rename.cc

```diff
@@ -0,0 +1,373 @@
+      .schedule_latency(schedule_latency)
+      .fetch_queues(&mock_L1I.queues)
+      .data_queues(&mock_L1D.queues)
+    };
+
+    uut.ROB.push_back(champsim::test::instruction_with_ip(1));
+    for (auto &instr : uut.ROB)
+      instr.ready_time = champsim::chrono::clock::time_point{};
+
+    constexpr int PHYSICALREGS = 128;
+    RegisterAllocator ra{PHYSICALREGS};
+
+    WHEN("A write and then a read on the same logical register are scheduled, but only the write executes"){
+      auto write1 = champsim::test::instruction_with_ip(1);
+      write1.destination_registers.push_back(5);
+      write1.instr_id = 1;
+      write1.destination_registers[0] = ra.rename_dest_register(write1.destination_registers[0], write1.instr_id);
+
+      auto read1 = champsim::test::instruction_with_ip(2);
+      read1.source_registers.push_back(5);
+      read1.instr_id = 2;
+      read1.source_registers[0] = ra.rename_src_register(read1.source_registers[0]);
+      
+      ra.complete_dest_register(write1.destination_registers[0]);
+
+      AND_WHEN("write1 retires"){
+        write1.completed = true;
+        ra.complete_dest_register(write1.destination_registers[0]);
+        ra.retire_dest_register(write1.destination_registers[0]);
+        THEN("No registers should have been recycled since no new writes to that arch reg"){
+          REQUIRE(ra.count_free_registers() == PHYSICALREGS-1);
+        }
+        AND_WHEN("the read instruction is retired"){
+          read1.completed = true;
+          THEN("No registers should have been recycled since no new writes to that arch reg"){
+            REQUIRE(ra.count_free_registers() == PHYSICALREGS-1);
+          }
+          AND_WHEN("A new write to the same register occurs"){
+            auto write2 = champsim::test::instruction_with_ip(3);
+            write2.destination_registers.push_back(5);
+            write2.instr_id = 3;
+            write2.destination_registers[0] = ra.rename_dest_register(write2.destination_registers[0], write2.instr_id);
+            THEN("there should be PHYSICALREGS-2 free registers"){
+              REQUIRE(ra.count_free_registers() == PHYSICALREGS-2);
+            }
+            AND_WHEN("The second write completes execution"){
+              ra.complete_dest_register(write2.destination_registers[0]);
+              THEN("There should be PHYSICALREGS-2 free registers because the instruction has not retired"){
+                REQUIRE(ra.count_free_registers() == PHYSICALREGS-2);
+              }
+              AND_WHEN("The second write has retired"){
+                write2.completed = true;
+                ra.retire_dest_register(write2.destination_registers[0]);
+                THEN("Exactly one physical register should have been freed."){
+                  REQUIRE(ra.count_free_registers() == PHYSICALREGS-1);
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+
+    WHEN("Ten writes to the same logical register occur"){
```

Comment:
I can't help but feel like there's a better way to write this. Even a loop would help comprehensibility here.